### PR TITLE
Reflection_oM: Adding PreviousVersion attribute for versioning support

### PR DESCRIPTION
**Pull request by adecler**
### Issues addressed by this PR
See https://github.com/BHoM/BHoM_UI/pull/247 for more details

Simply add a new attribute that will enable local versioning as an attribute directly applied to a method or property

### Test files
You will have to try you own here if you want to test. But you probably don't need to as it is just a very simple object definition.

### CODE REVIEW

**Comment by adecler** [comment]
Yes, we should add constructor to the list of authorised targets. I also think that the attribute should be named 'PreviousVersion' instead.

File: Reflection_oM/Attributes/OldVersionAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BH.oM.Reflection.Attributes
+{
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Property)]
+    public class OldVersionAttribute : Attribute, IImmutable
```
Comment (FraserGreenroyd) [inline Reflection_oM/Attributes/OldVersionAttribute.cs:34]: To be `PreviousVersion` as discussed

File: Reflection_oM/Attributes/OldVersionAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BH.oM.Reflection.Attributes
+{
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Property)]
```
Comment (FraserGreenroyd) [inline Reflection_oM/Attributes/OldVersionAttribute.cs:33]: To include `AttributeTargets.Constructor` as discussed

**Comment by adecler** [comment]
@FraserGreenroyd , I have made the changes discussed. I have also added `AllowMultiple = true` so we can keep all the previous version if necessary. My initial idea was to keep only the ones for the current version (so batch remove at the start of each quarter) but this might be a bit impractical and, even then, there is not guarantee that two methods would not be replaced by the same one.

File: Reflection_oM/Attributes/OldVersionAttribute.cs (hunk 1 above)
Comment (adecler) [reply Reflection_oM/Attributes/OldVersionAttribute.cs]: done

**Pull request merged**